cmd/HocusPocus.hs

```diff
@@ -42,33 +43,95 @@ optionsParser
 makeOptions headerText parser = Opt.info (Opt.helper <*> parser) (Opt.fullDesc <> Opt.header (toS headerText))
 
--- | Execute 'Command' against a Wormhole Rendezvous server.
-app :: Command -> Rendezvous.Session -> IO ()
-app command session = do
-  print command
+-- | A password used to exchange with a Magic Wormhole peer.
+--
+-- XXX: Just picking ByteString because that's the least amount of work. Need
+-- to look up exact type of password in the magic-wormhole docs.
+type Password = ByteString
+
+-- | Send a text message to a Magic Wormhole peer.
+sendText :: Rendezvous.Session -> Password -> Text -> IO ()
+sendText session password message = do
   nameplate <- Rendezvous.allocate session
   mailbox <- Rendezvous.claim session nameplate
   peer <- Rendezvous.open session mailbox  -- XXX: We should run `close` in the case of exceptions?
   let (Messages.Nameplate n) = nameplate
-  Peer.withEncryptedConnection peer (Spake2.makePassword (toS n <> "-potato"))
+  Peer.withEncryptedConnection peer (Spake2.makePassword (toS n <> "-" <> password))
     (\conn -> do
-        let offer = FileTransfer.Message "Brave new world that has such offers in it"
+        let offer = FileTransfer.Message message
         Peer.sendMessage conn (toS (Aeson.encode offer)))
 
+-- | Receive a text message from a Magic Wormhole peer.
+receiveText :: Rendezvous.Session -> IO Text
+receiveText session = do
+  nameplates <- Rendezvous.list session
+  putStrLn $ "Nameplates: " <> Text.intercalate ", " [ n | Messages.Nameplate n <- nameplates ]
+  putText "Choose nameplate: "
+  nameplate <- getLine
+  mailbox <- Rendezvous.claim session (Messages.Nameplate nameplate)
+  peer <- Rendezvous.open session mailbox
+  putText "Password: "
+  password <- getLine
+  let fullPassword = toS nameplate <> "-" <> toS password
+  Peer.withEncryptedConnection peer (Spake2.makePassword fullPassword)
+    (\conn -> do
+        received <- atomically $ Peer.receiveMessage conn
+        case Aeson.eitherDecode (toS received) of
+          Left err -> panic $ "Could not decode message: " <> show err
+          Right (FileTransfer.Message message) -> pure message)
+
+-- | Bounce a trivial message to and from a Rendezvous server.
+bounce :: WebSocketEndpoint -> Messages.AppID -> IO ()
+bounce endpoint appID = do
+  side1 <- Messages.generateSide
+  side2 <- Messages.generateSide
+  Rendezvous.runClient endpoint appID side1 $ \session1 -> do
+    nameplate <- Rendezvous.allocate session1
+    mailbox1 <- Rendezvous.claim session1 nameplate
+    peer1 <- Rendezvous.open session1 mailbox1
+    Rendezvous.runClient endpoint appID side2 $ \session2 -> do
+      mailbox2 <- Rendezvous.claim session2 nameplate
+      peer2 <- Rendezvous.open session2 mailbox2
+      let message = "aoeu"
+      (_, output) <- concurrently (send peer1 message) (receive peer2)
+      unless (output == message) $ panic $ "Mismatched messages: " <> show message <> " != " <> show output
```

Comment:
I read some blog post that talked about monads as a solution to the staircase pattern ... which this _almost_ is.  I wonder if you have any thoughts on that.

Reply:
Yes, I think I do!

The prolific, splendid, and friendly Gabriel Gonzales has made a package called [managed](https://hackage.haskell.org/package/managed-1.0.5/docs/Control-Monad-Managed.html) which deals with this situation almost exactly.

I highly recommend skimming through the introductory section of the documentation, because he says everything I could say in this comment but better.

Reply:
I'm not using managed here because I don't think the extra dependency is worth it for a small command-line app intended solely for testing. Wouldn't take much to change my mind though.

Reply:
Managed looks really cool.  Perhaps if/when the command-line app expands to include more functionality it will make sense to bring that in as a dependency.  I agree that there isn't really a problem here now that needs fixing.